File: examples/part5-testkit-go/testkit.go
// Package testkit is a tiny test framework — the tool this academy has been
// using, built from the inside.
//
// Three pieces every test framework shares:
//   - Assertions that turn a condition into a descriptive failure ("expected 4,
//     got 5"), not a bare false.
//   - A registry + runner that runs each test in isolation, so one test's
//     failure (or panic) does not stop the others.
//   - A canonical report — the "test x ... ok / FAILED" summary — deterministic
//     and byte-identical across languages.
//
// We test the framework the only honest way: feed it known-passing and
// known-failing tests and assert its report. No I/O — the report is a string.
package testkit

import "fmt"

// Check is the result of an assertion or test body: nil passed, a non-empty
// error failed with a human-readable reason.
type Check = error

// AssertTrue asserts a boolean condition, failing with message.
func AssertTrue(condition bool, message string) Check {
	if condition {
		return nil
	}
	return fmt.Errorf("%s", message)
}

// AssertEq asserts two values are equal, failing with a message showing both.
func AssertEq[T comparable](actual, expected T) Check {
	if actual == expected {
		return nil
	}
	return fmt.Errorf("expected %v, got %v", expected, actual)
}

type testCase struct {
	name string
	body func() Check
}

// TestKit is a registry of named tests. Register with Test, then Run.
type TestKit struct {
	tests []testCase
}

// New returns an empty kit.
func New() *TestKit { return &TestKit{} }

// Test registers a test under name. Chainable.
func (k *TestKit) Test(name string, body func() Check) *TestKit {
	k.tests = append(k.tests, testCase{name, body})
	return k
}

type outcome struct {
	name    string
	failed  bool
	message string
}

// Run runs every test in registration order, isolating each: a returned error
// is a failure, and a panic is recovered and turned into a failure too, so a
// crashing test cannot take down the run.
func (k *TestKit) Run() *Report {
	outcomes := make([]outcome, 0, len(k.tests))
	for _, tc := range k.tests {
		outcomes = append(outcomes, runOne(tc))
	}
	return &Report{outcomes: outcomes}
}

func runOne(tc testCase) (result outcome) {
	result.name = tc.name
	defer func() {
		if r := recover(); r != nil {
			result.failed = true
			result.message = fmt.Sprintf("panicked: %v", r)
		}
	}()
	if err := tc.body(); err != nil {
		result.failed = true
		result.message = err.Error()
	}
	return result
}

// Report is the result of a run: every outcome, plus a canonical text summary.
type Report struct {
	outcomes []outcome
}

// Passed returns the number of passing tests.
func (r *Report) Passed() int {
	n := 0
	for _, o := range r.outcomes {
		if !o.failed {
			n++
		}
	}
	return n
}

// Failed returns the number of failing tests.
func (r *Report) Failed() int { return len(r.outcomes) - r.Passed() }

// OK reports whether every test passed.
func (r *Report) OK() bool { return r.Failed() == 0 }

// Summary renders the canonical report — the exact format is the cross-language
// contract asserted by the tests.
func (r *Report) Summary() string {
	lines := []string{fmt.Sprintf("running %d tests", len(r.outcomes))}
	for _, o := range r.outcomes {
		status := "ok"
		if o.failed {
			status = "FAILED"
		}
		lines = append(lines, fmt.Sprintf("test %s ... %s", o.name, status))
	}
	lines = append(lines, "")

	var failures []outcome
	for _, o := range r.outcomes {
		if o.failed {
			failures = append(failures, o)
		}
	}
	if len(failures) > 0 {
		lines = append(lines, "failures:")
		for _, o := range failures {
			lines = append(lines, fmt.Sprintf("    %s: %s", o.name, o.message))
		}
		lines = append(lines, "")
	}

	result := "ok"
	if !r.OK() {
		result = "FAILED"
	}
	lines = append(lines, fmt.Sprintf("test result: %s. %d passed; %d failed",
		result, r.Passed(), r.Failed()))

	out := ""
	for i, l := range lines {
		if i > 0 {
			out += "\n"
		}
		out += l
	}
	return out
}

File: examples/part5-testkit-py/testkit.py
"""A tiny test framework — the tool this academy has been using, built from the
inside.

Three pieces every test framework (pytest, unittest, JUnit) shares:

* **Assertions** that turn a condition into a descriptive failure ("expected 4,
  got 5"), not a bare ``False``.
* **A registry + runner** that runs each test in **isolation**, so one test's
  failure (or exception) does not stop the others.
* **A canonical report** — the "test x ... ok / FAILED" summary — deterministic
  and byte-identical across languages.

We test the framework the only honest way: feed it known-passing and
known-failing tests and assert its report. No I/O — the report is a string.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass


class AssertionFail(Exception):
    """Raised by an assertion helper to signal a test failure with a message."""


def assert_true(condition: bool, message: str) -> None:
    """Assert a boolean condition, failing with ``message``."""
    if not condition:
        raise AssertionFail(message)


def assert_eq(actual: object, expected: object) -> None:
    """Assert two values are equal, failing with a message showing both."""
    if actual != expected:
        raise AssertionFail(f"expected {expected!r}, got {actual!r}")


@dataclass
class _Outcome:
    name: str
    failed: bool
    message: str


class TestKit:
    """A registry of named tests. Register with :meth:`test`, then
    :meth:`run`."""

    def __init__(self) -> None:
        self._tests: list[tuple[str, Callable[[], None]]] = []

    def test(self, name: str, body: Callable[[], None]) -> TestKit:
        """Register a test under ``name``. Chainable."""
        self._tests.append((name, body))
        return self

    def run(self) -> Report:
        """Run every test in registration order, isolating each: an
        :class:`AssertionFail` is a failure, and *any* other exception is caught
        and turned into a failure too, so a crashing test cannot take down the
        run."""
        outcomes: list[_Outcome] = []
        for name, body in self._tests:
            try:
                body()
                outcomes.append(_Outcome(name, False, ""))
            except AssertionFail as e:
                outcomes.append(_Outcome(name, True, str(e)))
            except Exception as e:  # noqa: BLE001 - test isolation is the point
                outcomes.append(_Outcome(name, True, f"raised: {e!r}"))
        return Report(outcomes)


class Report:
    """The result of a run: every outcome, plus a canonical text summary."""

    def __init__(self, outcomes: list[_Outcome]) -> None:
        self._outcomes = outcomes

    def passed(self) -> int:
        return sum(1 for o in self._outcomes if not o.failed)

    def failed(self) -> int:
        return sum(1 for o in self._outcomes if o.failed)

    def ok(self) -> bool:
        return self.failed() == 0

    def summary(self) -> str:
        """Render the canonical report — the exact format is the cross-language
        contract asserted by the tests."""
        lines = [f"running {len(self._outcomes)} tests"]
        for o in self._outcomes:
            status = "FAILED" if o.failed else "ok"
            lines.append(f"test {o.name} ... {status}")
        lines.append("")

        failures = [o for o in self._outcomes if o.failed]
        if failures:
            lines.append("failures:")
            for o in failures:
                lines.append(f"    {o.name}: {o.message}")
            lines.append("")

        result = "ok" if self.ok() else "FAILED"
        lines.append(
            f"test result: {result}. {self.passed()} passed; {self.failed()} failed"
        )
        return "\n".join(lines)


if __name__ == "__main__":
    import sys

    kit = (
        TestKit()
        .test("addition works", lambda: assert_eq(2 + 2, 4))
        .test("string upper", lambda: assert_eq("hi".upper(), "HI"))
        .test("deliberate failure", lambda: assert_eq(10 // 2, 4))
        .test("a truthy check", lambda: assert_true(3 > 1, "3 should be > 1"))
    )
    report = kit.run()
    print(report.summary())
    sys.exit(0 if report.ok() else 1)
